integration-tests/tests/src/tests/counter.ts

```diff
@@ -0,0 +1,994 @@
+        const objectWithMixed = this.realm.create<IWithMixed>(WithMixedSchema.name, { mixed: [20] });
+
+        return { counter, list, dictionary, set, objectWithMixed };
+      });
+      expectCounter(counter);
+      expectRealmList(list);
+      expectRealmDictionary(dictionary);
+      expectRealmSet(set);
+      expect(counter.value).equals(10);
+      expect(list[0]).equals(20);
+      expect(dictionary.key).equals(20);
+      expect(set[0]).equals(20);
+
+      // Collections OF Mixed
+      expect(() => {
+        this.realm.write(() => {
+          list[0] = counter;
+        });
+      }).to.throw("Using a Counter as a Mixed value is not supported");
+      expect(list[0]).equals(20);
+
+      expect(() => {
+        this.realm.write(() => {
+          list.push(counter);
+        });
+      }).to.throw("Using a Counter as a Mixed value is not supported");
+      expect(list.length).equals(1);
+      expect(list[0]).equals(20);
+
+      expect(() => {
+        this.realm.write(() => {
+          dictionary.key = counter;
+        });
+      }).to.throw("Using a Counter as a Mixed value is not supported");
+      expectKeys(dictionary, ["key"]);
+      expect(dictionary.key).equals(20);
+
+      expect(() => {
+        this.realm.write(() => {
+          dictionary.set({ newKey: counter });
+        });
+      }).to.throw("Using a Counter as a Mixed value is not supported");
+      expectKeys(dictionary, ["key"]);
+      expect(dictionary.key).equals(20);
+
+      expect(() => {
+        this.realm.write(() => {
+          set.add(counter);
+        });
+      }).to.throw("Using a Counter as a Mixed value is not supported");
+      expect(set[0]).equals(20);
+      expect(set.has(counter.value)).to.be.false;
+
+      // Collections IN Mixed
+      expect(() => {
+        this.realm.write(() => {
+          objectWithMixed.mixed = [counter];
+        });
+      }).to.throw("Using a Counter as a Mixed value is not supported");
+      expect(list.length).equals(1);
+      expect(list[0]).equals(20);
+
+      expect(() => {
+        this.realm.write(() => {
+          objectWithMixed.mixed = { newKey: counter };
+        });
+      }).to.throw("Using a Counter as a Mixed value is not supported");
+      expectKeys(dictionary, ["key"]);
+      expect(dictionary.key).equals(20);
+    });
+
+    it("throws when using counter as query argument", function (this: RealmContext) {
+      const { counter } = this.realm.write(() => {
+        return this.realm.create<IWithCounter>(WithCounterSchema.name, {
+          counter: 10,
+        });
+      });
+      expectCounter(counter);
+
+      const objects = this.realm.objects(WithCounterSchema.name);
+      expect(objects.length).equals(1);
+
+      expect(() => {
+        objects.filtered("counter = $0", counter);
+      }).to.throw("Using a Counter as a query argument is not supported. Use 'Counter.value'");
```

Comment:
It might be interesting to eventually lift this restriction too.

Reply:
Yup I agree. Went with our earlier agreement to push this to a future counter version.